traveltimepy/throttler.py

```diff
@@ -0,0 +1,98 @@
+import asyncio
+from collections import deque
+from contextlib import asynccontextmanager
+from loguru import logger
+
+
+class Throttler:
+    """Simple throttler for asyncio"""
+
+    def __init__(
+        self,
+        rate_limit: int,
+        time_window_seconds: int,
+        retry_interval=0.001
+    ):
+        self.time_window = time_window_seconds
+        self.rate_limit = rate_limit
+        self.retry_interval = retry_interval
+
+        # Set the event loop to the one passed in, or the default asyncio event loop if none is passed.
+        self.loop = asyncio.new_event_loop()
+        asyncio.set_event_loop(self.loop)
+
+        # Create a new deque object to store task start times.
+        self._task_logs = deque()
+
+    def remove_expired_tasks(self):
+        # Get the current time from the event loop
+        now = self.loop.time()
+
+        # Remove items (which are start times) that are no longer in the time window
+        # Check if the deque containing task start times is not empty.
+        while self._task_logs:
+            # Calculate the time difference between the current time and the first task start time in the deque,
+            # and check whether it's greater than the rate interval length.
+            if now - self._task_logs[0] > self.time_window:
+                # Remove the first(start time) item from the deque since it's no longer needed.
+                self._task_logs.popleft()
+            else:
+                # If the first item in the deque is still within the time window, break out of the while loop.
+                break
+
+    async def __aenter__(self):
+        await self.wait_for_availability()
+        return self
+
+    async def __aexit__(self, exc_type, exc_val, exc_tb):
+        # nothing to do here
+        pass
+
+    async def wait_for_availability(self, window_size: int):
```

Comment:
Let's say I add one request - it gets delayed for a second.
After 300 ms  free space opens up
After it opens up I add  second request (while the first one still sleeps)
So seconds one would go through while first one would still be asleep (if this process repeats 1st one in theory can be stuck).
Does this implementation prevent from this behaviour?

Reply:
yeap, it cleans out all requests which are done

Reply:
and free the space

Reply:
I'm talking about requests that are going to be completed/scheduled
In other words can described scenario happen when second request would get scheduled first due to timing

Reply:
well this was a theoretical question. In practice such problem probably won't happen unless you create some endless loop of requests that always reaches rate limit and  would get that that bad timing. If that was the case I would've suggested  this as a future improvement. But nvm then (unless someone else could be able to answer this question)

Reply:
I think you have a point here @MockusTravelTime if a task enters the throttler just in the time between the expiry of a previous task and a waiting task waking up:

To make it more clear:

1. Request A enters throttler, which sends it to sleep due to lack of capacity.
2. One of the older tasks expires
3. Before Task A wakes up another task B enters the throttler and passes immediately because now there's enough capacity
4. Task A wakes up

This scenario is made especially likely as `self.retry_interval` introduces a minimum wait time. Thinking it over again, I think the `retry_interval` is kind of obsolete as we calculate the wait time anyway. We should have either the retry_interval or the calculated wait time. My assumption was that a calculated wait time is better as in the case of the Travel Time API typical wait times will be in the range of multiple seconds, so we can avoid a lot of check loops vs. having a fixed retry_interval which would have to be a very short interval in that case.

That would however not 100% guarantee that requests are (potentially infinitely) pushed back. A solution would be:

Add each waiting request to a `queue` deck before sending it so sleep and add a check to line `61` that asserts that not only is there enough capacity but also that the current request is the next in line.

Reply:
@MxMartin if you want you can send us your solution of that changed throttler, so we'll be able to deploy it faster.